sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala

```diff
@@ -41,18 +46,114 @@ case class WriteToDataSourceV2(batchWrite: BatchWrite, query: LogicalPlan)
   override def output: Seq[Attribute] = Nil
 }
 
+case class AppendDataExec(
+    table: SupportsBatchWrite,
+    writeOptions: DataSourceOptions,
+    query: SparkPlan) extends V2TableWriteExec with BatchWriteHelper {
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    val batchWrite = newWriteBuilder() match {
+      case builder: SupportsSaveMode =>
+        builder.mode(SaveMode.Append).buildForBatch()
+
+      case builder =>
+        builder.buildForBatch()
+    }
+    doWrite(batchWrite)
+  }
+}
+
+case class OverwriteByExpressionExec(
+    table: SupportsBatchWrite,
+    filters: Array[Filter],
+    writeOptions: DataSourceOptions,
+    query: SparkPlan) extends V2TableWriteExec with BatchWriteHelper {
+
+  private def isTruncate(filters: Array[Filter]): Boolean = {
+    filters.length == 1 && filters(0).isInstanceOf[AlwaysTrue]
+  }
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    val batchWrite = newWriteBuilder() match {
+      case builder: SupportsTruncate if isTruncate(filters) =>
+        builder.truncate().buildForBatch()
+
+      case builder: SupportsOverwrite =>
+        builder.overwrite(filters).buildForBatch()
```

Comment:
This requires multiple test cases. 
- an empty filter list
- True
- False
- Null
- And/Or 
- some Filters that the data source does not support? e.g. `StringStartsWith`
- ...

Reply:
I agree that these tests should be added, but there isn't a good way to add them right now. My plan is to add these tests as functionality is exposed in improved APIs that make overwrite by expression available, along with a new test table implementation. Since this PR is already large and there is no way to access this through a public API, I think adding this can wait.